test/RoleTest.php

```diff
@@ -122,4 +122,66 @@ public function testCircleReferenceWithParent()
         $this->expectException(Exception\CircularReferenceException::class);
         $baz->addParent($foo);
     }
+
+    public function testGetPermissions()
+    {
+        $foo = new Role('foo');
+        $foo->addPermission('foo.permission');
+
+        $bar = new Role('bar');
+        $bar->addPermission('bar.permission');
+
+        $baz = new Role('baz');
+        $baz->addPermission('baz.permission');
+
+        // create hierarchy bar -> foo -> baz
+        $foo->addParent($bar);
+        $foo->addChild($baz);
+
+        $this->assertEquals([
+            'bar.permission',
+            'foo.permission',
+            'baz.permission'
+        ], $bar->getPermissions());
+        $this->assertEquals([
+            'bar.permission'
+        ], $bar->getPermissions(false));
+        $this->assertEquals([
+            'foo.permission',
+            'baz.permission'
+        ], $foo->getPermissions());
+        $this->assertEquals([
+            'foo.permission'
+        ], $foo->getPermissions(false));
+        $this->assertEquals([
+            'baz.permission'
+        ], $baz->getPermissions());
+        $this->assertEquals([
+            'baz.permission'
+        ], $baz->getPermissions(false));
```

Comment:
This test is not readable, maybe because of formating or maybe because we test here more than thing. Maybe we should use instead `dataProvider`? I think it should be clearer then.

Reply:
I think I can improve the readability adding some spaces between the assertions. Regarding the test scope I think it's fine, we are testing `getPermissions()` with an hierarchy model.

Reply:
I think I can improve the readability adding some spaces between the assertions. Regarding the test scope I think it's fine, we are testing getPermissions() with an hierarchy model.

Reply:
... or maybe just:

```php
        $this->assertEquals(['bar.permission', 'foo.permission', 'baz.permission'], $bar->getPermissions());
        $this->assertEquals(['bar.permission'], $bar->getPermissions(false));
        $this->assertEquals(['foo.permission', 'baz.permission'], $foo->getPermissions());
        $this->assertEquals(['foo.permission'], $foo->getPermissions(false));
        $this->assertEquals(['baz.permission'], $baz->getPermissions());
        $this->assertEquals(['baz.permission'], $baz->getPermissions(false));
```